Documentation/design-docs/host-testing.md

```diff
@@ -0,0 +1,100 @@
+# Testing of host components
+
+The hosting layer of .NET Core consists of several native libraries and executables (entirely written in C++). The full end to end features sometimes require code in the `coreclr` repo, but testing of that is outside of the code of this document. For description of the various hosting components, please see [host-components](host-components.md).
+
+Testing these comes with certain challenges which this document tries to address.
+
+## Existing tests
+Almost all tests for the hosting components are currently basically End-to-End tests which execute the entire product (with very limited amount of "mocking").
+
+### End-to-End tests
+This is almost all tests in the `HostActivation` test project currently.
+The tests prepare folders on the disk which emulate the product installation by having
+* The muxer `dotnet` or `apphost`
+* The `hostfxr` in the right place
+* Shared frameworks (or for self contained apps in the app)
+
+For the most part all these are real components. The shared frameworks contain a real `Microsoft.NETCore.App` copy, only higher level frameworks are "mocked" by creating only the necessary files.
+Almost all tests then execute a test application using this test product installation and observe the behavior of the app. This is done by
+* Having the app output information to standard output
+* Turning on host tracing and looking for certain text in the tracing output
+* Exit code of the app
+
+Pros:
+* True End-to-End tests which test real product binaries in the real-world conditions (mostly)
+* Tests are written in C# - so natural integration with test infra, VS UI and so on
+Cons:
+* All tests are out-of-proc which makes debugging product complicated
+* Tests perform large setup work and copy/write lot of files on disk - tests are slow for this reason.
+* Tests run the entire product, not just the host. This includes starting the runtime, JITing and so on. Vast majority of tests don't actually use that and effectively ignore the part of running an actual app code - makes the tests slow.
```

Comment:
How much of the slowdown is because of the file copies? and how much time is spent in the runtime, jit etc.? 
I'm wondering if its possible to ameliorate this situation by:
* Publishing a dummy JIT that does no work, and use it via `COMPLUS_ALTJIT`
* Using a dummy coreclr.dll that does no work

Reply:
For most tests dummy coreclr.dll would solve the problem.
We can definitely improve these tests - it would move us in the right direction (like mocking coreclr).